src/main/java/org/spongepowered/common/map/canvas/SpongeMapCanvasBuilder.java

```diff
@@ -0,0 +1,146 @@
+        return this;
+    }
+
+    @Override
+    public MapCanvas.Builder from(MapCanvas canvas) {
+        if (canvas instanceof SpongeEmptyCanvas) {
+            return MapCanvas.builder();
+        }
+        this.canvas = (SpongeMapByteCanvas)canvas;
+        return this;
+    }
+
+    @Override
+    public MapCanvas.Builder fromImage(Image image) {
+        Objects.requireNonNull(image, "image cannot be null");
+        if (image.getWidth(null) != 128 || image.getHeight(null) != 128) {
+            throw new IllegalStateException("image size was invalid!");
+        }
+        BufferedImage bufferedImage = null;
+        boolean shouldConvert = true;
+        if (image instanceof BufferedImage) {
+            bufferedImage = (BufferedImage)image;
+            // If its not TYPE_INT_RGB, convert it.
+            shouldConvert = bufferedImage.getType() != BufferedImage.TYPE_INT_RGB;
+        }
+        if (shouldConvert) {
+            bufferedImage = new BufferedImage(image.getWidth(null), image.getHeight(null), BufferedImage.TYPE_INT_RGB);
+            Graphics2D bGr = bufferedImage.createGraphics();
+            bGr.drawImage(image, 0, 0, null);
+            bGr.dispose();
+        }
+        int[] pixels = ((DataBufferInt) bufferedImage.getRaster().getDataBuffer()).getData();
+        // Get the color palette we are working with
+        Map<Integer, SpongeMapColor> palette = new HashMap<>();
+        for (MapColorType type : Sponge.getRegistry().getAllOf(MapColorType.class)) {
+            // Put each shade in also.
+            for (int i = 0; i < Constants.Map.MAP_SHADES; i++) {
+                SpongeMapColor spongeMapColor = new SpongeMapColor(type, i);
+                palette.put(spongeMapColor.getColor().getRgb(), spongeMapColor);
+            }
+        }
+        if (this.canvas == null) {
+            this.canvas = new SpongeMapByteCanvas();
+        }
+        for (int i = 0; i < pixels.length; i++) {
+            SpongeMapColor color = palette.get(pixels[i]);
+            if (color == null) {
+                throw new IllegalStateException("Can not find a matching color for rgb value: " + pixels[i] + ". The MapCanvas will have painted all pixels up to this point.");
+            }
+            this.canvas.setPixel(i, color.getMCColor());
+        }
+        return this;
+    }
+
+    @Override
+    protected Optional<MapCanvas> buildContent(DataView container) throws InvalidDataException {
+        if (!container.contains(DataQuery.of("MapColors"))) {
+            return Optional.empty();
+        }
+        SpongeMapCanvas mapCanvas = new SpongeMapByteCanvas(
+                Bytes.toArray(container.getByteList(DataQuery.of("MapColors")).get())
+        );
+        return Optional.of(mapCanvas);
+    }
+
+    @Override
+    public MapCanvas build() {
+        if (this.canvas == null) {
+            return SpongeEmptyCanvas.INSTANCE;
+        }
+        return this.canvas;
```

Comment:
This is problematic.

Imagine that I create a canvas using this builder, then build it. You return the backing canvas here. Now I change something using this builder - that's going to change the already built canvas.

**The builder and provided object should have no relationship after the object is built**. The canvas currently has various paint methods, they really should be here on this builder, then a deep copy of the final data should be placed onto the canvas. The provided canvas should be immutable.

Please separate the canvas and the builder, only creating an immutable cavas when you build.